lang/src/org/partiql/lang/typesystem/interfaces/function/SqlFunction.kt

```diff
@@ -0,0 +1,42 @@
+package org.partiql.lang.typesystem.interfaces.function
+
+import org.partiql.lang.eval.ExprValue
+import org.partiql.lang.typesystem.interfaces.type.Type
+import org.partiql.lang.typesystem.interfaces.type.TypeWithParameters
+
+/**
+ * Used to define a sql function.
+ */
+interface SqlFunction {
+    /**
+     * Function name
+     */
+    fun getFuncName(): String
+
+    /**
+     * Required arguments of this function. Empty list means no required arguments.
+     */
+    fun getRequiredArgTypes(): List<Type>
+
+    /**
+     * Optional arguments of this function. Empty list means no optional arguments.
+     */
+    fun getOptionalArgTypes(): List<Type>
+
+    /**
+     * Variadic argument of this function. Null value means no variadic argument.
+     */
+    fun getVariadicArgType(): Type?
+
+    /**
+     * Function return type inference
+     */
+    fun inferReturnType(): List<TypeWithParameters>
```

Comment:
Just looking at this function... if "infer return type" means what I think it means, this function also needs an argument `List<Type>` here, since one of the inputs to inference will be the arguments of a call site.

Reply:
The argument types are already known by `requiredArgTypes`, `optionalArgTypes` and `variadicArgType`, right?

Reply:
Can you be more clear about why we need the argument of `List<Type>` here?

Reply:
I spent some time thinking about this, and I don't think we really need it. We can just add a property `returnType` for sql functions.